format/index.md

```diff
@@ -1004,125 +790,176 @@ Content-Type: application/vnd.api+json
     }
   }
 }
 ```
 
-In order to remove a to-one relationship, specify `null` as the value of the
-relationship.
+A server **MAY** reject an attempt to do a full replacement of a to-many
+relationship. In such a case, the server **MUST** reject the entire update,
+and return a `403 Forbidden` response.
 
-Alternatively, a to-one relationship **MAY** be accessible at its relationship
-URL (see above).
+> Note: Since full replacement may be a very dangerous operation, a server
+may choose to disallow it. A server may reject full replacement if it has
+not provided the client with the full list of associated objects, and does
+not want to allow deletion of records the client has not seen.
 
-A `PUT` request sent to the URL of a relationship **SHOULD** update the
-relationship. For example:
+#### Responses <a href="#crud-updating-responses" id="crud-updating-responses" class="headerlink"></a>
 
-```text
-PUT /articles/1/links/author
-Content-Type: application/vnd.api+json
-Accept: application/vnd.api+json
+##### 204 No Content <a href="#crud-updating-responses-204" id="crud-updating-responses-204" class="headerlink"></a>
 
-{
-  "people": "12"
-}
-```
+A server **MUST** return a `204 No Content` status code if an update is
+successful and the client's current attributes remain up to date.
 
-A `PUT` request **SHOULD** succeed regardless of whether a relationship is
-currently defined.
+##### 200 OK <a href="#crud-updating-responses-200" id="crud-updating-responses-200" class="headerlink"></a>
 
-A to-one relationship **MAY** alternatively be added by sending a `POST`
-request with an individual primary resource to the URL of the relationship.
-For example:
+If a server accepts an update but also changes the resource(s) in other ways
+than those specified by the request (for example, updating the `updatedAt`
+attribute or a computed `sha`), it **MUST** return a `200 OK` response.
 
-```text
-POST /articles/1/links/author
-Content-Type: application/vnd.api+json
-Accept: application/vnd.api+json
+The response document for a `200 OK` **MUST** include a representation of
+the updated resource(s) as if a `GET` request was made to the request URL.
 
-{
-  "people": "12"
-}
-```
+##### 403 Forbidden <a href="#crud-updating-relationship-responses-403" id="crud-updating-relationship-responses-403" class="headerlink"></a>
 
-A `POST` request should only succeed if no relationship is currently defined.
+A server **MUST** return `403 Forbidden` in response to an unsupported request
+to update a resource or relationship.
 
-A to-one relationship **MAY** be removed by sending a `DELETE` request to
-the URL of the relationship. For example:
+##### 404 Not Found <a href="#crud-updating-responses-404" id="crud-updating-responses-404" class="headerlink"></a>
 
-```text
-DELETE /articles/1/links/author
-```
+A server **MUST** return `404 Not Found` when processing a request to modify
+a resource that does not exist.
 
```

Comment:
DELETE is an [idempotent method](http://tools.ietf.org/html/rfc7231#section-4.2.2)

[DELETE method](http://tools.ietf.org/html/rfc7231#section-4.3.5)
```
If a DELETE method is successfully applied, the origin server SHOULD
   send a 202 (Accepted) status code if the action will likely succeed
   but has not yet been enacted, a 204 (No Content) status code if the
   action has been enacted and no further information is to be supplied,
   or a 200 (OK) status code if the action has been enacted and the
   response message includes a representation describing the status.
```

One can argue that because DELETE is idempotent then a 204 response for an already deleted resource is most appropriate if it is known that the resource did exist at some point. If it is not known then an error response seems most appropriate.

HTTP response codes are designed to allow the server to be as specific as it can with errors partitioned into classes.

The error classes that seem most appropriate in order of specificity are:

[Response 410](http://tools.ietf.org/html/rfc7231#section-6.5.9)

[Response 404](http://tools.ietf.org/html/rfc7231#section-6.5.4)

[Response 5xx](http://tools.ietf.org/html/rfc7231#section-6.6):
```   The 5xx (Server Error) class of status code indicates that the server
   is aware that it has erred or is incapable of performing the
   requested method.
```
Given the above definition, 500 should be seen as the error of last resort as it reflects **the server** has *erred* or cannot perform the operation (implying it ordinarily would if it was able to). It certainly does not help the client or intermediate proxy/caches to know the state of the resource.

Maybe a better approach is some guidance around the priority of appropriate responses anchored to the RFC.  It would need to cover the the use cases of deleting a resource that once existed, vs deleting a resource that never existed, vs deleting a resource where previous existence is not known.